### PR TITLE
Implement String.prototype.substring()

### PR DESCRIPTION
**Pull request by lvidacs**
JerryScript-DCO-1.0-Signed-off-by: Laszlo Vidacs lvidacs.u-szeged@partner.samsung.com

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp [hunk 1]
```diff
@@ -421,7 +421,130 @@ ecma_builtin_string_prototype_object_substring (ecma_value_t this_arg, /**< this
                                                 ecma_value_t arg1, /**< routine's first argument */
                                                 ecma_value_t arg2) /**< routine's second argument */
 {
-  ECMA_BUILTIN_CP_UNIMPLEMENTED (this_arg, arg1, arg2);
+  ecma_completion_value_t ret_value = ecma_make_empty_completion_value ();
+
+  /* 1 */
+  ECMA_TRY_CATCH (check_coercible_val,
+                  ecma_op_check_object_coercible (this_arg),
+                  ret_value);
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp]: Is this implicitly changes the this value?

File: tests/jerry/string-prototype-substring.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+assert("hello world!".substring(5, obj) === " world!");
+
+// check other objects
+var obj = { substring : String.prototype.substring }
+
+obj.toString = function() {
+    return "Iam";
+}
+assert(obj.substring(100000,1) === "am");
```
Comment (zherczeg) [inline tests/jerry/string-prototype-substring.js]: we should try assert(true.subString() === ); and similar objectsd. I don't know if coercible works.

File: jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp [hunk 3]
```diff
@@ -421,7 +421,130 @@ ecma_builtin_string_prototype_object_substring (ecma_value_t this_arg, /**< this
...
+
+    /* 9 */
+    uint32_t to = start > end ? start : end;
+
+    /* 10 */
+    /* Workaround: avoid repeated call of ecma_string_get_char_at_pos() because its overhead */
+    uint32_t zt_str_size = (uint32_t) sizeof (ecma_char_t) * (len + 1);
+    ecma_char_t *original_zt_str_p = (ecma_char_t*) mem_heap_alloc_block (zt_str_size,
+                                                                          MEM_HEAP_ALLOC_SHORT_TERM);
+    ecma_string_to_zt_string (original_string_p, original_zt_str_p, (ssize_t) zt_str_size);
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp]: I think this is changed now to some UTF-8 processing.

**Comment by lvidacs** [comment]
Continued in PR #318

**Pull request closed without merging**